torch/utils/module_tracker.py

```diff
@@ -74,6 +93,54 @@ def is_bw(self):
         """
         return torch._C._current_graph_task_id() != -1
 
+    def get_known_fqn(self, mod):
+        """
+        Return the fqn for the given module if it is known to the Module Tracker, otherwise None
+        """
+
+        if mod not in self._known_modules:
+            return None
+        return self._known_modules[mod]
+
+    def register_user_hooks(
+        self,
+        pre_fw_hook: Optional[Callable] = None,
+        post_fw_hook: Optional[Callable] = None,
+        pre_bw_hook: Optional[Callable] = None,
+        post_bw_hook: Optional[Callable] = None,
+    ):
+        """
+        Registers user specified hooks that are called before/after the forward/backward pass for
+        each module tracked by the Module Tracker. One or more can be none.
+
+        The hooks should have the following signatures.
+
+        pre_fw_hook (fqn, module, input) : Called before the forward pass for the module.
+        post_fw_hook (fqn, module, input, output) : Called after the forward pass for the module.
+        pre_bw_hook (fqn, *outputs) : A multi-grad hook that is called on all the outputs of the
```

Comment:
Can we make the signature consistent here and always take input/output as objects and not varargs?

Reply:
The backward hooks are multi-grad hooks registered only on the inputs/outputs that require grad, hence we need varargs for the backward hooks. But this is not a problem for the forward hooks. And they are consistent with the signature of pre/post forward hooks and multi-grad hooks respectively.

Reply:
Why do you need varargs? The tracker can get them as varargs from the autograd engine and pass them as a tuple to the hook

Reply:
Okay, got it.